### ISSUE TITLE
IonLoading pattern usage

### ISSUE THREAD

**Issue by rclai** [meteoric/meteor-ionic#21]
I'm trying to implement this routing pattern:

```
Router.route('/myroute', {
  ... other route stuff ...
  onBeforeAction: function () {
    IonLoading.show({ delay: 0 });
    this.next();
  },
  onAfterAction: function () {
    IonLoading.hide();
  }
})
```

But I'm getting this error:

```
Exception in callback of async function: TypeError: Cannot read property 'firstNode' of undefined
    at Object.IonLoading.hide
```

This probably happens because the firstNode hasn't finished rendering when onAfterAction gets triggered.

Is there a pattern you guys follow to get that loading screen to pop up when there's a route with say, a big subscription?

Also, can you guys refactor so that we don't have to send an object into the show method? That way we can just do IonLoading.show(). Currently, doing that will throw an error because the delay property is not found.

**Comment by nickw** [comment]
Hmm yeah I haven't tried to do this yet. I think because of the way `iron:router` works, you'll have to configure a loading template, and then perhaps do your `IonLoading` calls in the `rendered` and `destroyed` callbacks of that template.

> Also, can you guys refactor so that we don't have to send an object into the show method? That way we can just do IonLoading.show(). Currently, doing that will throw an error because the delay property is not found.

Yeah, good call. Can you open a second issue for that?

**Comment by rclai** [comment]
The loading template using rendered and destroyed worked great.

**Comment by benstr** [comment]
@rclai I am glad you have it fixed. @nickw I followed your suggestion and I still get the loader stuck on the page.

**Controller**

```
ProjectsController = AppController.extend({
  subscriptions: function() {
    return Meteor.subscribe('gcProjectsMyIndex');
  },
  data: function() {
    return {
      projectsIndex: Projects.find({},{sort:{jobNumber:-1}})
    };
  },
  action: function() {
    if (this.ready()) {
      this.render('projectsIndex');
    } else {
      this.render('loading');
    }
  }
});
```

**Loading HTML**

```
<template name="loading">
</template>
```

**Loading Logic**

```
Template.loading.rendered = function() {
  IonLoading.show({delay: 50});
};
Template.loading.destroyed = function() {
  IonLoading.hide();
};
```

**Comment by rclai** [comment]
In subscriptions, aren't you supposed to do ```this.subscribe()```?

**Comment by benstr** [comment]
@rclai I do not see it, am I missing somthing?
https://github.com/EventedMind/iron-router/blob/devel/Guide.md#the-subscriptions-option

**Comment by AnkitTheCoder** [comment]
I am seeing the same issue as @benstr. While using the iron-router's default loading hook, the app never goes to the index page, but rather gets stuck on loading template. The only difference in terms of the code, from what Ben has posted, is that I am not explicitly using the `this.ready()` call. I am relying on iron-router calling the loading template, since my templates uses `waitOn`. 

I too am using rendered and destroyed within the loading template. I am seeing this more prominently when using trying on mobile app (using `meteor run ios-device` command).

Any suggestions/ideas on how to fix this?

**Comment by nickw** [comment]
I've realized that using `iron:router`'s `waitOn` and `data` features don't fit well with the way mobile routes are supposed to animate. `iron:router` takes a very traditional, server-side approach to rendering templates (wait until all the data is loaded, then render), whereas mobile apps and modern web apps take a more fine-grained approach and render pieces as needed (see [Sacha's blog post on template-level subscriptions](https://www.discovermeteor.com/blog/template-level-subscriptions/)).

I've implemented a pattern in the meteoric example apps (see [this commit](https://github.com/meteoric/contacts/commit/9de81a7a1e69bfdbfc78f78609b00dc6bb155b5c#diff-9cdca4bffae39727221ae7143159000fR1) in the contacts app) that works much better by allowing the template to be rendered an animated to immediately (no choppiness) while showing a loading spinner and loading the data on demand. 

Once template-level subscriptions are supported in the Meteor core, I can remove a lot of the boilerplate from this pattern, but in the meantime I wouldn't recommend using `waitOn` with `meteor-ionic`.
